File: lib/templates/layouts/page.ts
import { Site } from "../../site"
import { html } from ".."
import { layout } from "./default"

const _page = (title: string, content: string) => html`
  <h1>${title}</h1>
  ${content}
`

export const page = (site: Site, title: string, content: string): string => {
  return layout(site, title, _page(title, content))
}

File: app/Templates/Page.hs
{-# LANGUAGE ExtendedDefaultRules #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE OverloadedStrings #-}

module Templates.Page
  ( pageTemplate,
  )
where

import Lucid
import Templates.Footer (footerScript, footerTemplate)
import Templates.Head (headTemplate)
import Templates.Header (headerTemplate)
import Types (Content (getContent, getTitle), SiteMeta)

mainBody :: Content a => a -> Html ()
mainBody content =
  main_
    [class_ "main"]
    ( do
        h1_ (toHtml $ getTitle content)
        (toHtml $ getContent content)
    )

pageTemplate :: Content a => SiteMeta -> a -> Html ()
pageTemplate meta content =
  doctype_
    *> ( html_
           [lang_ "en", data_ "theme" "light"]
           ( do
               headTemplate meta content
               body_
                 [class_ "root"]
                 ( do
                     headerTemplate
                     mainBody content
                     footerTemplate
                     footerScript meta
                 )
           )
       )
